internal/app/machined/pkg/controllers/network/device_config.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1"
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+	"github.com/talos-systems/talos/pkg/machinery/resources/network"
+)
+
+// DeviceConfigController manages network.DeviceConfig based on configuration.
+type DeviceConfigController struct {
+	devices map[string]networkDevice
+}
+
+type networkDevice struct {
+	hardwareAddress string
+	busPrefix       string
+	driver          string
+	pciID           string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *DeviceConfigController) Name() string {
+	return "network.DeviceConfigController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *DeviceConfigController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        pointer.To(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.LinkStatusType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *DeviceConfigController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.DeviceConfigSpecType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *DeviceConfigController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	ctrl.devices = map[string]networkDevice{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		links, err := r.List(ctx, resource.NewMetadata(network.NamespaceName, network.LinkStatusType, "", resource.VersionUndefined))
+		if err != nil {
+			return err
+		}
+
+		touchedIDs := make(map[resource.ID]struct{})
+
+		var cfgProvider talosconfig.Provider
+
+		cfg, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, config.MachineConfigType, config.V1Alpha1ID, resource.VersionUndefined))
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting config: %w", err)
+			}
+		} else {
+			cfgProvider = cfg.(*config.MachineConfig).Config()
+		}
+
+		if cfgProvider != nil {
+			selectedInterfaces := map[string]struct{}{}
+
+			for _, device := range cfgProvider.Machine().Network().Devices() {
+				if device.Selector() != nil {
+					err = ctrl.getDeviceBySelector(ctx, device, links.Items)
+					if err != nil {
+						logger.Warn("failed to select an interface for a device", zap.Error(err))
+
+						continue
+					}
+
+					if _, ok := selectedInterfaces[device.Interface()]; ok {
+						return fmt.Errorf("the device %s is already configured by a selector", device.Interface())
+					}
+
+					selectedInterfaces[device.Interface()] = struct{}{}
+				}
+
+				touchedIDs[device.Interface()] = struct{}{}
+
+				config := network.NewDeviceConfig(device)
```

Comment:
this looks like a breaking change - today Talos allows configuring same interface several times, e.g.:

```
- interface: eth0
  mtu: 1452
- interface: eth0
  addresses: [192.168.0.1/24]
```

As `DeviceConfig` has interface name as primary key, this is going to be a problem